### PR TITLE
Update the LOR-to-HO interface

### PR DESCRIPTION
**Pull request by v-dobrev**
See the new abstract base class `GridTransfer` and the derived classes `InterpolationGridTransfer` and `L2ProjectionGridTransfer`.

### CODE REVIEW

File: fem/fespace.hpp [hunk 1]
```diff
@@ -609,44 +597,206 @@ class QuadratureSpace
...
+    reference space and do not depend on the actual physical shape of the mesh
+    elements.
+
+    It is assumed that both the coarse and the fine FiniteElementSpace%s use
+    compatible types of elements, e.g. finite elements with the same map-type
+    (VALUE, INTEGRAL, H_DIV, H_CURL - see class FiniteElement). Generally, the
+    FE spaces can have different orders, however, in order for the backward
+    operator to be well-defined, the (local) number of the fine dofs should not
+    be smaller than the number of coarse dofs. */
+class InterpolationGridTransfer : public GridTransfer
```
Comment (tzanio) [inline fem/fespace.hpp]: Very nice description, thanks Veselin!

**Comment by tzanio** [comment]
I like these changes, including the `GridTransfer` name.

File: fem/fespace.hpp [hunk 2]
```diff
@@ -609,44 +597,206 @@ class QuadratureSpace
...
-    element space on a coarse mesh, and a low-order L2 finite element space on a
-    refined mesh (LOR). We assume that the low-order space, fes_lor, lives on a
-    mesh obtained by refining the mesh of the high-order space, fes_ho. */
-class L2Projection : public Operator
+
+/** @brief Base class for transfer algorithms that construct transfer Operator%s
+    between two finite element (FE) spaces. */
+/** Generally, the two FE spaces (domain and range) can be defined on different
+    meshes. */
+class GridTransfer
```
[the remark this answers is not in the document]
Comment (pazner) [reply fem/fespace.hpp]: My understanding is that `GridTransfer` interfaces could be much more general data/solution transfer, but right now the two implementations we have (`L2ProjectionGridTransfer` and `InterpolationGridTransfer`) are specific to the refinement case.

File: fem/fespace.hpp [hunk 3]
```diff
@@ -609,44 +597,206 @@ class QuadratureSpace
...
+/** The forward, coarse-to-fine, transfer uses L2 projection. The backward,
+    fine-to-coarse, transfer is defined locally (on a coarse element) as
+    B = (F^t M_f F)^{-1} F^t M_f, where F is the forward transfer matrix, and
+    M_f is the mass matrix on the union of all fine elements comprising the
+    coarse element. Note that the backward transfer operator, B, is a left
+    inverse of the forward transfer operator, F, i.e. B F = I. Both F and B are
+    defined in physical space and, generally, vary between different mesh
+    elements.
+
+    This class currently supports only L2 finite element spaces and fine meshes
```
Comment (pazner) [inline fem/fespace.hpp]: The L2 transfer operator will "work" even on H1 finite element spaces (it just won't have the desired conservation properties). It won't completely break though.

**Comment by pazner** [comment]
The new interface looks good to me. Thanks @v-dobrev for your changes/fixed to the L2 version!

File: fem/fespace.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (v-dobrev) [reply fem/fespace.hpp]: Yes, in the base class, we do not need to make an assumption. The two derived classes specify that as a requirement (the "range" is a refinement of the "domain") in the class description.

File: fem/fespace.hpp (hunk 3 above)
Comment (v-dobrev) [reply fem/fespace.hpp]: Oh, okay, I did not realize that. Please update the documentation to reflect that.

Comment (pazner) [reply fem/fespace.hpp]: Ok, I added some explanatory comments.

**Pull request merged**